src/Evolution/DgSubcell/PerssonTci.cpp

```diff
@@ -0,0 +1,76 @@
+// Distributed under the MIT License.
+// See LICENSE.txt for details.
+
+#include "Evolution/DgSubcell/PerssonTci.hpp"
+
+#include <cstddef>
+
+#include "DataStructures/ApplyMatrices.hpp"
+#include "DataStructures/Index.hpp"
+#include "DataStructures/Matrix.hpp"
+#include "NumericalAlgorithms/Spectral/Filtering.hpp"
+#include "NumericalAlgorithms/Spectral/Mesh.hpp"
+#include "Utilities/GenerateInstantiations.hpp"
+#include "Utilities/Gsl.hpp"
+#include "Utilities/TMPL.hpp"
+
+namespace evolution::dg::subcell::detail {
+template <>
+double average_order(const Index<1>& extents) noexcept {
+  return static_cast<double>(extents[0] - 1);
+}
+
+template <>
+double average_order(const Index<2>& extents) noexcept {
+  return static_cast<double>(extents[0] + extents[1] - 2) * 0.5;
+}
+
+template <>
+double average_order(const Index<3>& extents) noexcept {
+  return static_cast<double>(extents[0] + extents[1] + extents[2] - 3) *
+         0.333333333333333;
+}
+
+template <size_t Dim>
+bool persson_tci_impl(const gsl::not_null<DataVector*> filtered_component,
+                      const DataVector& component, const Mesh<Dim>& dg_mesh,
+                      const double avg_order, const double alpha) noexcept {
+  ASSERT(component.size() == dg_mesh.number_of_grid_points(),
+         "The tensor components being checked must have the same number of "
+         "grid points as the DG mesh. The tensor has "
+             << component.size() << " grid points while the DG mesh has "
+             << dg_mesh.number_of_grid_points() << " grid points.");
+
+  const Matrix identity{};
+  for (size_t d = 0; d < Dim; ++d) {
+    auto matrices = make_array<Dim>(std::cref(identity));
+    gsl::at(matrices, d) = Spectral::filtering::zero_lowest_modes(
+        dg_mesh.slice_through(d), dg_mesh.extents(d) - 1);
+    apply_matrices(filtered_component, matrices, component, dg_mesh.extents());
+
+    // Avoid taking logs of small numbers (or worse, zero)
+    if (l2Norm(*filtered_component) / l2Norm(component) < 1.0e-18) {
```

Comment:
so I've been undecided on whether to hard-code the `1.0e-18` epsilon, and I'm leaning more and more to that being a bad idea. I think passing in an `epsilon` to avoid logs of small numbers would be a good thing. Thoughts @kidder @fmahebert ?

Related: this should also be doing `(l2norm(filt) < epsilon * l2norm(component))` to avoid division by zero